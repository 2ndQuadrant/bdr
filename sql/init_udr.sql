\c postgres
SELECT bdr.bdr_subscribe(
	local_node_name := 'postgres',
	remote_dsn := 'dbname=regression',
	local_dsn := 'dbname=postgres',
	replication_sets := ARRAY['default', 'important', 'for-node-2', 'for-node-2-insert', 'for-node-2-update', 'for-node-2-delete']
	);

SELECT bdr.bdr_node_join_wait_for_ready();

-- Make sure we see the slot and active connection
SELECT plugin, slot_type, database, active FROM pg_replication_slots;
SELECT count(*) FROM pg_stat_replication;

SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections;
SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes;

\c regression
SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections;
SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes;

-- emulate the pg_xlog_wait_remote_apply on vanilla postgres
DO $DO$BEGIN
	PERFORM 1 FROM pg_proc WHERE proname = 'pg_xlog_wait_remote_apply';
	IF FOUND THEN
		RETURN;
	END IF;

	PERFORM bdr.bdr_replicate_ddl_command($DDL$
		CREATE OR REPLACE FUNCTION public.pg_xlog_wait_remote_apply(i_pos pg_lsn, i_pid integer) RETURNS VOID
		AS $FUNC$
		BEGIN
			WHILE EXISTS(SELECT true FROM pg_stat_get_wal_senders() s WHERE s.flush_location < i_pos AND (i_pid = 0 OR s.pid = i_pid)) LOOP
				PERFORM pg_sleep(0.01);
			END LOOP;
		END;$FUNC$ LANGUAGE plpgsql;
	$DDL$);
END;$DO$;

SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE OR REPLACE FUNCTION public.bdr_regress_variables(
    OUT readdb1 text,
    OUT readdb2 text,
    OUT writedb1 text,
    OUT writedb2 text
    ) RETURNS record LANGUAGE SQL AS $f$
SELECT
    current_setting('bdrtest.readdb1'),
    current_setting('bdrtest.readdb2'),
    current_setting('bdrtest.writedb1'),
    current_setting('bdrtest.writedb2')
$f$;
$DDL$);
