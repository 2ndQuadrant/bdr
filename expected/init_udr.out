\c postgres
SELECT bdr.bdr_subscribe(
	local_node_name := 'node-pg',
	subscribe_to_dsn := 'dbname=regression',
	node_local_dsn := 'dbname=postgres',
	replication_sets := ARRAY['default', 'important', 'for-node-2', 'for-node-2-insert', 'for-node-2-update', 'for-node-2-delete']
	);
 bdr_subscribe 
---------------
 
(1 row)

SELECT bdr.bdr_node_join_wait_for_ready();
 bdr_node_join_wait_for_ready 
------------------------------
 
(1 row)

-- Make sure we see the slot and active connection
SELECT plugin, slot_type, database, active FROM pg_replication_slots;
 plugin | slot_type |  database  | active 
--------+-----------+------------+--------
 bdr    | logical   | regression | t
(1 row)

SELECT count(*) FROM pg_stat_replication;
 count 
-------
     1
(1 row)

SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections;
     conn_dsn      |                                conn_replication_sets                                 
-------------------+--------------------------------------------------------------------------------------
 dbname=regression | {default,important,for-node-2,for-node-2-insert,for-node-2-update,for-node-2-delete}
(1 row)

SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes;
 node_status | node_local_dsn  | node_init_from_dsn 
-------------+-----------------+--------------------
 r           |                 | 
 r           | dbname=postgres | dbname=regression
(2 rows)

\c regression
SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections;
 conn_dsn | conn_replication_sets 
----------+-----------------------
(0 rows)

SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes;
 node_status | node_local_dsn | node_init_from_dsn 
-------------+----------------+--------------------
(0 rows)

-- emulate the pg_xlog_wait_remote_apply on vanilla postgres
DO $DO$BEGIN
	PERFORM 1 FROM pg_proc WHERE proname = 'pg_xlog_wait_remote_apply';
	IF FOUND THEN
		RETURN;
	END IF;

	PERFORM bdr.bdr_replicate_ddl_command($DDL$
		CREATE OR REPLACE FUNCTION public.pg_xlog_wait_remote_apply(i_pos pg_lsn, i_pid integer) RETURNS VOID
		AS $FUNC$
		BEGIN
			WHILE EXISTS(SELECT true FROM pg_stat_get_wal_senders() s WHERE s.flush_location < i_pos AND (i_pid = 0 OR s.pid = i_pid)) LOOP
				PERFORM pg_sleep(0.01);
			END LOOP;
		END;$FUNC$ LANGUAGE plpgsql;
	$DDL$);
END;$DO$;
SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE OR REPLACE FUNCTION public.bdr_regress_variables(
    OUT readdb1 text,
    OUT readdb2 text,
    OUT writedb1 text,
    OUT writedb2 text
    ) RETURNS record LANGUAGE SQL AS $f$
SELECT
    current_setting('bdrtest.readdb1'),
    current_setting('bdrtest.readdb2'),
    current_setting('bdrtest.writedb1'),
    current_setting('bdrtest.writedb2')
$f$;
$DDL$);
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

